### PR TITLE
Update to use CUDA

### PR DESCRIPTION
**Pull request by stefanozampini**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Things to do:
* [ ] how to enable CUDA 12?

### CODE REVIEW

**Comment by stefanozampini** [comment]
Hypre complex failures should be fixed by https://gitlab.com/petsc/petsc/-/merge_requests/7497

**Comment by stefanozampini** [comment]
Should the tests be disabled when `libcuda.so` is not present? https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=921270&view=logs&jobId=696704cc-6fef-57a3-ea36-f27779b8cd5e&j=2ed0550f-e734-5ebf-52c6-1bfc19f5811c&t=9c969b19-a14a-5482-8fb9-78bdf745c3c6

**Comment by dalcinl** [comment]
> Should the tests be disabled when libcuda.so is not present?

Maybe there is a way to link with `libcuda.so` from the stubs, perhaps via setting `LIBRARY_PATH` (for the GCC+linker)  and `LD_LIBRARY_PATH` (for the dynamic linker) ? Or there are not even stubs to search for?

**Comment by stefanozampini** [comment]
Maybe add __cuda as the set of requirements for the test?

**Comment by dalcinl** [comment]
> . I think we should use `runpath`

I'm not sure what do you mean exactly. Are you talking about passing `-Wl,--enable-new-dtags` to use new newer `RUNPATH` semantics instead of the older `RPATH` semantics in shared libraries? In general, I would agree, although with the particular of CUDA, I'm not fully aware of all the gory details.

**Comment by stefanozampini** [comment]
@dalcinl Builds are clean, I'm skipping running the compiled tests for `aarch64` for now. We can figure out the cross-compilation issue with curand later

**Comment by dalcinl** [comment]
Stefano, if you know for a fact a build is broken and will not work, I believe the usual conda-forge approach to it is to skip it, simply because broken package files should not be published. Or this is just a CI testing issue, and the packages to work if installed in systems with newer enough GLIBC? Have you tried? In that case, then yes, we can move forward. Please confirm (or we can discuss it in person in a couple hours).

**Comment by dalcinl** [comment]
@stefanozampini Just in case, can do it with two commits? One for the recipe/config changes, and another for the re-rendering changes from conda-smithy? Maybe this is not really necessary, but that's how I've always done it, and I prefer to follow established patterns in things I do not understand in full :sweat_smile:.

**Comment by dalcinl** [comment]
@leofang Any chance you can give a quick glance to the changes to `meta.yml` here?

**Review by dalcinl — APPROVED** [review]

**Comment by leofang** [comment]
You have CUDA 12 support added to your recipe, but no CI/CD pipelines spawned. I suggest that you manually add [the CUDA 12 migrator](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/migrations/cuda120.yaml) to `.ci_support/migrations` and rerender.

For complex recipes like this one, cupy-feedstock can be a good reference (for supporting both CUDA 11/12 which have different file layouts + many CUDA packages + Python + ...), but no big issue from a quick glance other than the missing CIs.

**Comment by dalcinl** [comment]
> For complex recipes like this one, cupy-feedstock can be a good reference

We did, but we were not sure about the CUDA 12 migration, so @stefanozampini leaved it out. Now we know better.
Many thanks, @leofang !

**Comment by leofang** [comment]
No worries, last time I heard we're preparing a doc refresher for conda-forge CUDA packages. It might take some time to finalize, sorry for the learning curve. Before it lands, feel free to ping me or anyone from `@conda-forge/cuda` to ask for help.

**Comment by stefanozampini** [comment]
@leofang, I see cupy feedstock only builds with 12.4 https://github.com/conda-forge/cupy-feedstock/blob/main/recipe/meta.yaml#L37. Is the idea to use  11.8 or 12.4 only?

**Comment by stefanozampini** [comment]
thanks. So, to support the same in PETSc, I need to copy-paste the ignore_run_exports paraphernalia and related run_constrained stuff that are in cupy?

File: recipe/meta.yaml [hunk 1]
```diff
@@ -110,3 +172,4 @@ extra:
     - SylvainCorlay
     - martinRenou
     - matthiasdiener
+    - stefanozampini
```
Comment (dalcinl) [inline recipe/meta.yaml:175]: And hell will unravel over your head...

**Pull request merged**